pageserver/src/tenant.rs

```diff
@@ -2194,28 +2194,44 @@ impl Tenant {
         let (state, mut rx) = watch::channel(state);
 
         tokio::spawn(async move {
-            let mut current_state: &'static str = From::from(&*rx.borrow_and_update());
             let tid = tenant_id.to_string();
-            TENANT_STATE_METRIC
-                .with_label_values(&[&tid, current_state])
-                .inc();
+
+            fn inspect_state(state: &TenantState) -> ([&'static str; 1], bool) {
+                ([state.into()], matches!(state, TenantState::Broken { .. }))
+            }
+
+            let mut tuple = inspect_state(&rx.borrow_and_update());
+
+            let is_broken = tuple.1;
+            if !is_broken {
+                // the tenant might be ignored and reloaded, so first remove any previous set
+                // element. it most likely has already been scraped, as these are manual operations
+                // right now. most likely we will add it back very soon.
+                drop(crate::metrics::BROKEN_TENANTS_SET.remove_label_values(&[&tid]));
+            }
+
             loop {
-                match rx.changed().await {
-                    Ok(()) => {
-                        let new_state: &'static str = From::from(&*rx.borrow_and_update());
-                        TENANT_STATE_METRIC
-                            .with_label_values(&[&tid, current_state])
-                            .dec();
-                        TENANT_STATE_METRIC
-                            .with_label_values(&[&tid, new_state])
-                            .inc();
-
-                        current_state = new_state;
-                    }
-                    Err(_sender_dropped_error) => {
-                        info!("Tenant dropped the state updates sender, quitting waiting for tenant state change");
-                        return;
-                    }
+                let labels = &tuple.0;
+                let current = TENANT_STATE_METRIC.with_label_values(labels);
+                current.inc();
+
+                if rx.changed().await.is_err() {
+                    // tenant has been dropped; decrement the counter because a tenant with that
+                    // state is no longer in tenant map, but allow any broken set item to exist
+                    // still.
+                    current.dec();
+                    break;
+                }
```

Comment:
Paraphrasing what I read here, please confirm my understanding. You may dismiss or apply the suggestion.

```suggestion
                if rx.changed().await.is_err() {
                    // tenant has been dropped; decrement the counter because a tenant with that
                    // state is no longer in tenant map; the metric with the tenant_id label
                    // will remain, but its value is 0. we can't remove the metric with
                    // the tenant_id label because it could race with re-attach of same tenant.
                    current.dec();
                    break;
                }
```

Reply:
The value in set is left as 1, not decremented to zero.

The `current.dec()` representing the participation in count of "currently active/broken/stopping/whatever" tenants is decremented. That is a separate thing. 

The re-attach does not even matter here; the drop will happen before the next attach unless someone holds an Arc to bully. I want these semantics of "leaking" the broken value even while tenant has been ignored, as I added in the test. At least I don't know we would have any short-lived broken tenants which disappear. If we would have, this might make them visible.

Reply:
Ok, got it, let's disregard attach.

And it doesn't matter that this task can outlive the tenant object because the new task will `drop(crate::metrics::BROKEN_TENANTS_SET.remove_label_values(&[&tid]));` , thereby redirecting the old task's `current` into the void.

Very clever

Reply:
No current is not the `BROKEN_TENANTS_SET`. It is the per-state counter for Active and other tenant states.

Reply:
So, the idea is to never decrement the per-tenant-id broken counter?
How's that going to work in the dashboard?
Alert on per-state gauge, but show table of per-tenant-id-broken-counter increases in last X hours?

Reply:
I am sure there's some promql way to get the tenants out, it's not that much different from the previous version. Please notice I am doing my best to keep the "good stuff" from previous version while limiting the issues it had (listing this set membership for each tenant).

Reply:
Yeah, I was trying to argue that "the good stuff" from the previous version maybe wasn't super useful in the first place, so, I would be OK with cutting it down if it speeds things up.